internal/pppoe/session.go

```diff
@@ -26,17 +28,52 @@ func connectSessionFd(fd int, ifName string, remote net.HardwareAddr, sessionID
 	return unix.Connect(fd, sa)
 }
 
-func sendSessionPacket(fd int, pkt []byte) error {
-	n, err := unix.Write(fd, pkt)
+func sendSessionPacket(fd int, pkt []byte, deadline time.Time) (n int, err error) {
+	if !deadline.IsZero() {
+		if err = setTimeout(fd, unix.SO_SNDTIMEO, deadline); err != nil {
+			return 0, err
+		}
+		defer func() {
+			resetErr := setTimeout(fd, unix.SO_SNDTIMEO, time.Time{})
+			if err == nil && resetErr != nil {
+				err = resetErr
+			}
+		}()
+	}
+	n, err = unix.Write(fd, pkt)
 	if err != nil {
-		return err
+		return n, err
 	}
 	if n != len(pkt) {
-		return fmt.Errorf("short socket write: got %d, want %d", n, len(pkt))
+		return n, fmt.Errorf("short socket write: got %d, want %d", n, len(pkt))
 	}
-	return nil
+	return n, nil
 }
 
-func readSessionPacket(fd int, buf []byte) (n int, from unix.Sockaddr, err error) {
+func readSessionPacket(fd int, buf []byte, deadline time.Time) (n int, from unix.Sockaddr, err error) {
+	if !deadline.IsZero() {
+		if err = setTimeout(fd, unix.SO_RCVTIMEO, deadline); err != nil {
+			return 0, nil, err
+		}
+		defer func() {
+			resetErr := setTimeout(fd, unix.SO_RCVTIMEO, time.Time{})
+			if err == nil && resetErr != nil {
+				err = resetErr
+			}
+		}()
+	}
 	return unix.Recvfrom(fd, buf, 0)
 }
+
+func setTimeout(fd int, opt int, deadline time.Time) error {
+	var tv unix.Timeval
+	if !deadline.IsZero() {
+		d := time.Until(deadline).Truncate(time.Microsecond)
+		if d < 0 {
+			return errors.New("timeout") // TODO: net.Error
```

Comment:
easy enough I'd do now. For instance, x/net/http2 contains:

```go
type httpError struct {
        msg     string
        timeout bool
}

func (e *httpError) Error() string   { return e.msg }
func (e *httpError) Timeout() bool   { return e.timeout }
func (e *httpError) Temporary() bool { return true }
```

Make something like that, or copy that.

Reply:
Done.